server/src/main/java/org/elasticsearch/search/rank/feature/RankFeatureShardPhase.java

```diff
@@ -48,10 +53,31 @@ public static void prepareForFetch(SearchContext searchContext, RankFeatureShard
 
         RankFeaturePhaseRankShardContext rankFeaturePhaseRankShardContext = shardContext(searchContext);
         if (rankFeaturePhaseRankShardContext != null) {
-            assert rankFeaturePhaseRankShardContext.getField() != null : "field must not be null";
-            searchContext.fetchFieldsContext(
-                new FetchFieldsContext(Collections.singletonList(new FieldAndFormat(rankFeaturePhaseRankShardContext.getField(), null)))
-            );
+            String field = rankFeaturePhaseRankShardContext.getField();
+            assert field != null : "field must not be null";
+            searchContext.fetchFieldsContext(new FetchFieldsContext(Collections.singletonList(new FieldAndFormat(field, null))));
+            CustomRankInput customRankInput = rankFeaturePhaseRankShardContext.getCustomRankInput();
+            if (customRankInput instanceof SnippetRankInput snippetRankInput && snippetRankInput.snippets() != null) {
```

Comment:
This logic should live in the text similarity rank builder imo. See my previous comment, we don't need to pass the custom rank input, we can let the rank builder modifies the search context to setup the information it needs from the fetch phase and then let them materialise in the search hits?